src/Groups/directproducts.jl

```diff
@@ -0,0 +1,486 @@
+   Gfull=DirectProductOfGroups(G.Xfull,G.L,G.n,G.Xfull,true)
+   @assert length(V)==G.n "Vector of wrong length"
+   x = prod([embedding(Gfull,j)(V[j]) for j in 1:length(V)])
+   @assert x in G "Element not in the group"
+   return group_element(G,x.X)
+end
+
+# start part on subgroups
+function _as_subgroup_bare(G::DirectProductOfGroups, H::GapObj)
+  t = H==G.X
+  return DirectProductOfGroups(H, G.L, G.n, G.X, t)
+end
+
+function _as_subgroup(G::DirectProductOfGroups, H::GapObj, ::Type{U}) where U
+  H1 = _as_subgroup_bare(G, H)
+  return H1, hom(H1, G, x::U -> group_element(G, x.X))
+end
+
+function _as_subgroup(G::DirectProductOfGroups, H::GapObj)
+  return _as_subgroup(G, H, GAPGroupElem{DirectProductOfGroups})
+end
+
+function sub(G::DirectProductOfGroups, elms::Vector{GAPGroupElem{DirectProductOfGroups}})
+  elems_in_GAP = GAP.julia_to_gap(GapObj[x.X for x in elms])
+  H = GAP.Globals.Group(elems_in_GAP)
+  #H is the group. I need to return the inclusion map too
+  return _as_subgroup(G, H)
+end
+
+function sub(L::GAPGroupElem{DirectProductOfGroups}...)
+   if length(L)==0 throw(ArgumentError("Empty list")) end
+   l=collect(L)
+   @assert all(x -> parent(x) == parent(l[1]), l)
+   return sub(parent(l[1]),l)
+end
+
+function sub(L::Vector{GAPGroupElem{DirectProductOfGroups}})
+   if length(L)==0 throw(ArgumentError("Empty list")) end
+   l=collect(L)
+   @assert all(x -> parent(x) == parent(l[1]), l)
+   return sub(parent(l[1]),l)
+end
+
+function Base.show(io::IO, G::DirectProductOfGroups)
+   if G.isfull
+      print(io, "DirectProduct of ")
+      display(G.L)
+   else
+      print(io, GAP.gap_to_julia(GAP.Globals.StringViewObj(G.X)))
+   end
+end
+
+#=
+function Base.show(io::IO, x::GAPGroupElem{DirectProductOfGroups{S,T}}) where { S, T }
+   print(io, "DirectProductElement( ", GAP.gap_to_julia(GAP.Globals.StringViewObj(projection(parent(x),1)(x).X)), " , ", GAP.gap_to_julia(GAP.Globals.StringViewObj(projection(parent(x),2)(x).X)), " )")
+end
+=#
+
+# if a subgroup of a direct product of groups is also a direct product of groups
+"""
+    write_as_full(G::DirectProductOfGroups)
+If `G` is a subgroup of the direct product `G_1 x ... x G_n` such that `G = H_1 x ... x H_n` for `H_i` subgroup of `G_i`, return `G` as full direct product of the `H_i`. If such `H_i` do not exist, an ERROR is returned.
+"""
+function write_as_full(G::DirectProductOfGroups)
+   if G.isfull
+      return G
+   else
+      LK = [image(projection(G,j))[1] for j in 1:G.n]
+      H = direct_product(LK)
+# index(H,G)==1 does not work because it does not recognize G as a subgroup of H
+      if order(H)==order(G)
+         return H
+      else
+         throw(ArgumentError("G is not a direct product of groups"))
+      end
+   end
+end
+
+"""
+    isfull_direct_product(G::DirectProductOfGroups)
+Return whether `G` is direct product of its factors (`false` if it is a proper subgroup).
+"""
+isfull_direct_product(G::DirectProductOfGroups) = G.isfull
```

Comment:
If somebody really asks for it, I am fine with computing it; but does this really justify precomputing `isfull` in all cases, even when we never ask for it? I don't think so...

Reply:
The field `isfull` is used in some functions like `embedding` and `projection`. For example, the projection is always defined, but for some reason GAP allows it only for full direct products, so in the other cases it has to be implemented. Without the parameter `isfull`, we would need to make a computation similar to the one in the function `write_as_full` every time. This could be more expensive. And this was also the reason why I put the check `t = H==G.X` in the function `_as_subgroup_bare`; it could be expensive in the first moment, but it could save time later.

Reply:
> but for some reason GAP allows it only for full direct products
??? I have no idea what you mean there? Perhaps you tried to call `Projection` and `Embedding` on subgroups of a direct product? Yes, that won't work, you always have to call them on the "full" group -- but you are storing this anyway (resp. it is available via `parent`), so where is the problem?

I really don't understand why you think you'd need to compute "fullness" every time. Perhaps show me the code you think you'd have to write, then we can discuss alternatives. As it is, I feel we are talking past each other

Reply:
The "fullness" is required for all of these functions that are not definable on proper subgroups, such as `embedding` and `projection`, or the function `(G::DirectProductOfGroups)(V::GAPGroupElem...)`, that is made to write explicitly an element in a direct product. In these function I do the computation in the Oscar full direct product (the `parent`) and then move on the subgroup. Do you mean that I can do the same process directly with GAP functions and variable (so, without produce the full direct product in Oscar) and transpose everything in Oscar only at the end?
For example, substituting the `projection` function by the following
```suggestion
function projection(G::DirectProductOfGroups, j::Base.Integer)
   f=GAP.Globals.Projection(G.Xfull,j)
   j in 1:length(G.L) || throw(ArgumentError("index not valid"))
   H = G.L[j]
   p = GAP.Globals.GroupHomomorphismByFunction(G.X,H.X,y->GAP.Globals.Image(f,y))
   return _hom_from_gap_map(G,H,p)
end
```